### ISSUE TITLE
Make additional attributes splittable in `RandomLinkSplit`

### ISSUE THREAD

**Issue by aridf** [pyg-team/pytorch_geometric#6006]
### 🚀 The feature, motivation and pitch

I am trying to perform edge classification on a dataset like this:

```
HeteroData(
  artist={ node_id=[80382] },
  release_group={ node_id=[352997] },
  (artist, collaborated_on, release_group)={
    edge_index=[2, 3323975],
    edge_label=[3323975],
    fuzzy_mask=[3323975]
  },
  (release_group, rev_collaborated_on, artist)={
    edge_index=[2, 3323975],
    edge_label=[3323975],
    fuzzy_mask=[3323975]
  }
)
```
Where `edge_label` is the label to be predicted and `fuzzy_mask` is a boolean tensor indicating the nodes that should be sampled during training / testing. I still want the full graph accessible for neighbour sampling.

When I split the data I get a train set like this:
```
transform = T.RandomLinkSplit(
        num_val = 0.1,
        num_test = 0.1,
        # Dynamically splitting original and reversed edge types
        edge_types = [f for f in data.metadata()[1] if not "rev_" in f[1]],
        rev_edge_types = [f for f in data.metadata()[1] if "rev_" in f[1]],
        # Don't want any negative samples for this task
        neg_sampling_ratio=0.0, 
        # Instead, the key is edge_label
        key = 'edge_label',
        is_undirected = True
    )
train_data, val_data, test_data = transform(data)
train_data

HeteroData(
  artist={ node_id=[80382] },
  release_group={ node_id=[352997] },
  (artist, collaborated_on, release_group)={
    edge_index=[2, 2659181],
    fuzzy_mask=[2659181],
    edge_label=[2659181],
    edge_label_index=[2, 2659181]
  },
  (release_group, rev_collaborated_on, artist)={
    edge_index=[2, 2659181],
    edge_label=[2659181],
    fuzzy_mask=[2659181]
  }
)
```
and a val set like this  
```
HeteroData(
  artist={ node_id=[80382] },
  release_group={ node_id=[352997] },
  (artist, collaborated_on, release_group)={
    edge_index=[2, 2659181],
    fuzzy_mask=[2659181],
    edge_label=[332397],
    edge_label_index=[2, 332397]
  },
  (release_group, rev_collaborated_on, artist)={
    edge_index=[2, 2659181],
    edge_label=[2659181],
    fuzzy_mask=[2659181]
  }
)
```
First of all, can you clarify why the dimensions of `edge_label_index` and `edge_index` differ in the validation data but not the training data? I think it's because we need the larger `edge_index` to access neighbors for aggregation, but then it's unclear to me why at training the two fields would have the same dimensions.

Second, I now have a problem. I want to create a loader that will only iterate over edges that are in `edge_label_index` but that also have `true` in the `fuzzy_mask`, but `fuzzy_mask` wasn't split by `RandomLinkSplit` so the two now have different sizes / do not line up.

Is there an easy way I could have avoided this problem? It does seem like the `edge_label_index` parameter in `LinkNeighborLoader` is meant to enable more precise sampling during training, but it seems quite cumbersome to implement given a mask. It would be nice if I `RandomLinkSplit` had an option to split additional edge attributes along with `edge_label_index`. Does this make sense? 

### Alternatives

I tried a function like this for constructing train and validation loaders:

```
def make_loader(data, fuzzy_subset = True):
    
    if fuzzy_subset:
        np_edge_index = np.array(data['artist', 'collaborated_on', 'release_group'].edge_index)
        sample_edges = np_edge_index[:,np.where(data['artist', 'collaborated_on', 'release_group'].fuzzy_mask == 1)]
        edge_label_index = torch.tensor(sample_edges).squeeze()
         
    else:
        edge_label_index = data['artist', 'collaborated_on', 'release_group'].edge_label_index
    
    edge_label = data['artist', 'collaborated_on', 'release_group'].edge_label

    
    print("Dim of edge labels to sample")
    print(edge_label_index.size())

    loader = LinkNeighborLoader(
            data=data,  
            num_neighbors=[10, 5] * 2, 
            neg_sampling_ratio=0, 
            edge_label_index=(('artist', 'collaborated_on', 'release_group'), edge_label_index),
            edge_label=edge_label,
            batch_size=2000,
            shuffle=True,
        )

    return loader
```

But noticed that when I use this function for the train and val data, the val data ends up iterating over `2659181` edges instead of `332397`. 

### Additional context

If it's relevant, here is the model

```
class GNN(torch.nn.Module):
    def __init__(self, hidden_channels):
        super().__init__()

        self.conv1 = SAGEConv((-1, -1), hidden_channels)
        self.conv2 = SAGEConv((-1, -1), hidden_channels)

    def forward(self, x: Tensor, edge_index: Tensor) -> Tensor:
        x = self.conv1(x, edge_index).relu()
        x = self.conv2(x, edge_index)
        return x

class Classifier(torch.nn.Module):
    def forward(self, artist: Tensor, rg: Tensor, edge_label_index: Tensor) -> Tensor:
        # Convert node embeddings to edge-level representations:
        edge_feat_artist = artist[edge_label_index[0]]
        edge_feat_rg = rg[edge_label_index[1]]

        # Apply dot-product to get a prediction per supervision edge:
        return (edge_feat_artist * edge_feat_rg).sum(dim=-1)

class Model(torch.nn.Module):
    def __init__(self, full_dataset, hidden_channels):
        super().__init__()
        # Since the dataset does not come with rich features, we also learn embeddings for each etype
        # Embeddings are defined dynamically and stored in a dict to permit multiple etypes
        for etype in full_dataset.metadata()[0]:
            setattr(self, f'emb_{etype}', torch.nn.Embedding(full_dataset[etype].num_nodes, hidden_channels))

        # Instantiate homogeneous GNN:
        self.gnn = GNN(hidden_channels)

        # Convert GNN model into a heterogeneous variant:
        self.gnn = to_hetero(self.gnn, metadata=full_dataset.metadata())

        self.classifier = Classifier()

    def forward(self, data: HeteroData) -> Tensor:
        x_dict = dict()
        for etype in data.metadata()[0]:
           x_dict[etype] = getattr(self, f'emb_{etype}')(data[etype].node_id)

        # `x_dict` holds feature matrices of all node types
        # `edge_index_dict` holds all edge indices of all edge types
        x_dict = self.gnn(x_dict, data.edge_index_dict)

        pred = self.classifier(
            x_dict["artist"],
            x_dict["release_group"],
            data["artist", "collaborated_on", "release_group"].edge_label_index,
        )

        return pred
```

And here is how I'm currently splitting the data:

```
def split_data(data):
    transform = T.RandomLinkSplit(
        num_val = 0.1,
        num_test = 0.1,
        # Dynamically splitting original and reversed edge types
        edge_types = [f for f in data.metadata()[1] if not "rev_" in f[1]],
        rev_edge_types = [f for f in data.metadata()[1] if "rev_" in f[1]],
        # Don't want any negative samples for this task
        neg_sampling_ratio=0.0, 
        # Instead, the key is edge_label
        key = 'edge_label',
        is_undirected = True
    )
    train, val, test = transform(data)
    return train, val, test
```

**Comment by aridf** [comment]
I would still advocate for a feature that lets users split attributes to match `edge_label_index` instead of `edge_index`. 

I did my own implementation by first constructing a dictionary that maps edges to their mask, then looping over `edge_label_index` and only keeping ones that pass the mask. Code is below for others / for future reference.

```
def make_loader(data, fuzzy_masks = None, batch_size=20000):
    
    if fuzzy_masks:
        final_edge_index = []
        final_labels = []
        for ii in tqdm(np.arange(data['artist', 'collaborated_on', 'release_group'].edge_label_index.size()[1])):
            edge = data['artist', 'collaborated_on', 'release_group'].edge_label_index[:,ii]
            if fuzzy_masks[tuple(edge.tolist())] == 1:
                final_edge_index.append(tuple(edge.tolist()))
                final_labels.append(data['artist', 'collaborated_on', 'release_group'].edge_label[ii])
        final_edge_index = torch.tensor(list(zip(final_edge_index))).squeeze().T
        final_labels = torch.tensor(final_labels)
         
    else:
        final_edge_index = data['artist', 'collaborated_on', 'release_group'].edge_label_index
        final_labels = data['artist', 'collaborated_on', 'release_group'].edge_label

    
    print("Dim of edge labels to sample")
    print(final_edge_index.size())

    loader = LinkNeighborLoader(
            data=data,  
            num_neighbors=[10, 5] * 2, 
            neg_sampling_ratio=0, 
            edge_label_index=(('artist', 'collaborated_on', 'release_group'), final_edge_index),
            edge_label=final_labels,
            batch_size=batch_size,
            shuffle=True,
        )

    return loader
```
But I find no matches. Same holds if I look at

**Comment by rusty1s** [comment]
Thanks for sharing your code, glad you are unblocked.
